middleware/oapi_validate.go

```diff
@@ -75,67 +90,60 @@ func OapiRequestValidatorWithOptions(swagger *openapi3.T, options *Options) func
 			next.ServeHTTP(w, r)
 		})
 	}
-
 }
 
 // This function is called from the middleware above and actually does the work
 // of validating a request.
-func validateRequest(r *http.Request, router routers.Router, options *Options) (int, error) {
-
-	// Find route
-	route, pathParams, err := router.FindRoute(r)
+func validateOAPIRequest(r *http.Request, options Options) (int, error) {
+	// pain
+	route, pathParams, err := options.router.FindRoute(r)
 	if err != nil {
-		return http.StatusBadRequest, err // We failed to find a matching route for the request.
+		return http.StatusBadRequest, err
 	}
 
 	// Validate request
-	requestValidationInput := &openapi3filter.RequestValidationInput{
+	reqValidation := &openapi3filter.RequestValidationInput{
 		Request:    r,
 		PathParams: pathParams,
 		Route:      route,
-	}
-
-	if options != nil {
-		requestValidationInput.Options = &options.Options
+		Options:    options.Options,
 	}
 
 	// Validate security before any other validation, unless options.Options.MultiError is true
-	if options == nil || !options.Options.MultiError {
-		if err := validateSecurity(requestValidationInput); err != nil {
+	if options.Options == nil || !options.Options.MultiError {
+		security := reqValidation.Route.Operation.Security
+		if security == nil {
+			security = &reqValidation.Route.Spec.Security
+		}
+
+		if err := openapi3filter.ValidateSecurityRequirements(r.Context(), reqValidation, *security); err != nil {
 			return http.StatusUnauthorized, err
 		}
 	}
 
 	// Validate the rest of the request
-	if err := openapi3filter.ValidateRequest(context.Background(), requestValidationInput); err != nil {
-		switch e := err.(type) {
-		case *openapi3filter.RequestError:
+	if err := openapi3filter.ValidateRequest(r.Context(), reqValidation); err != nil {
+		reqError := &openapi3filter.RequestError{}
+		secError := &openapi3filter.SecurityRequirementsError{}
+		otherErrr := &openapi3.MultiError{}
+
+		switch {
+		case errors.As(err, &reqError):
 			// We've got a bad request
 			// Split up the verbose error by lines and return the first one
 			// openapi errors seem to be multi-line with a decent message on the first
-			errorLines := strings.Split(e.Error(), "\n")
+			errorLines := strings.Split(err.Error(), "\n")
 			return http.StatusBadRequest, fmt.Errorf(errorLines[0])
-		case *openapi3filter.SecurityRequirementsError:
+		case errors.As(err, &secError):
 			return http.StatusUnauthorized, err
-		default:
+		case errors.As(err, &otherErrr):
```

Comment:
Does this work as-is?
`reqError := &openapi3filter.RequestError{}`
`errors.As(..., &reqError)`

Reply:
Yes, for some reason we need double pointer to errors, so I need to instantiate them somewhere, I can't think of anything else that would avoid it.

Reply:
Oh I see, openapi is returning pointer errors.